### DISCUSSION TITLE
Create a custom validation attribute

### DISCUSSION THREAD

**Discussion by savr-reda** [spatie/laravel-data#542] [General]
Hi,

The [current documentation](https://spatie.be/docs/laravel-data/v3/advanced-usage/validation-attributes#content-creating-your-validation-attribute) isn't very explicit regarding how to create custom validation attributes. I need to create a custom validation attribute that I could apply to my DTO properties, however, I could not find a proper way to achieve this.

I came across several issues and questions and tried several implementations ([here](https://github.com/spatie/laravel-data/discussions/263) and [here](https://github.com/spatie/laravel-data/discussions/389) for example), but there's no explicit answer to how to achieve something like this:

```php
class Dog extends Data
{
    #[AllowedName] // Validation attribute with custom logic
    public string $name;
}
```

```php
#[Attribute(Attribute::TARGET_PROPERTY)]
class AllowedName extends ValidationRule
{
    public function getRules(): array
    {
       // AllowedNameValidationRule implements Illuminate\Contracts\Validation\ValidationRule
        return [new AllowedNameValidationRule()];
    }
}
```

However I get:

```php
PHP Error:  Object of class Rules\AllowedName could not be converted to string in /var/www/html/vendor/laravel/framework/src/Illuminate/Validation/ValidationRuleParser.php on line 134
```

Can you please provide some more explanation?

**Comment by rubenvanassche** [comment]
Yeah this was a bit difficult. I've added a new class + updated the docs on this, release normally later today.

**Comment by clementbirkle** [comment]
After searching for a while, I found the solution: the problem occurs when the rule doesn't implement the `ValidationRule` contract.

Here is an example of how to fix that:

```
<?php

namespace App\Rules;

use Attribute;
use Illuminate\Contracts\Validation\ValidationRule
use Spatie\LaravelData\Support\Validation\ValidationPath;
use Spatie\LaravelData\Attributes\Validation\CustomValidationAttribute;

#[Attribute(Attribute::TARGET_PROPERTY | Attribute::TARGET_PARAMETER)]
class CustomRule extends CustomValidationAttribute implements ValidationRule
{
    /**
     * @return array<object|string>|object|string
     */
    public function getRules(ValidationPath $path): array|object|string;
    {
        return new CustomRule();
    }
}
```
